hts.c

```diff
@@ -424,6 +424,18 @@ htsFile *hts_open_format(const char *fn, const char *mode, const htsFormat *fmt)
         *mode_c = format_to_mode[fmt->format];
     }
 
+    // If we really asked for a compressed text format but didn't
+    // specify 'z' format,  add it anyway.
+    if (strchr(mode, 'w') && fmt && fmt->compression == bgzf) {
+        if (fmt->format == sam || fmt->format == vcf || fmt->format == text_format) {
+            char *cp;
+            if ((cp = strchr(smode, 'b')))
```

Comment:
Hmm, staring at the contorted logic a bit longer I think you're right.  It's not obvious at first glance though. :-)

Thanks.

Reply:
Unsure whether to also add in tests above:

```
	else if (*cp == 'z')
            fmt_code = 'z';
```

That would mean hts_open_format with mode "wz" an fmt as SAM with no_compression set wouldn't write compressed data.  However I'm unsure if we care.  It follows the logic of the format taking precedence I guess so it's done, but I can see maybe people wanting an explicit compression in the mode string to be honoured for ease of coding (why add it otherwise)?